### PR TITLE
Add "semi-fungibles" support

### PR DESCRIPTION
**Pull request by jordy25519**
Adds:
- Support for SFTs
Changes:
- collection schema replaced by one-off series attributes
- Listings are keyed by listing IDs
- Rename create -> mint
- Tokens get global IDs `token_id = (collection_id, series_id, serial_number)`
- Add batch transfer/create functions
- Add mint additional for increasing supply of tokens in a series
- disabled individual token royalties as it's at odds with bundle sales. #441 would be necessary to resolve properly

partially addresses #437

### CODE REVIEW

**Comment by jordy25519** [comment]
run benchmark

File: crml/nft/src/lib.rs [hunk 1]
```diff
@@ -204,6 +220,7 @@ decl_module! {
 
 		/// Check and close all expired listings
 		fn on_initialize(now: T::BlockNumber) -> Weight {
+			// TODO: this is unbounded and could become costly
```
Comment (aliXsed) [inline crml/nft/src/lib.rs:239]: Can you please create an issue (if not already) and add the issue number here?

File: crml/nft/src/lib.rs [hunk 2]
```diff
@@ -157,24 +156,33 @@ decl_storage! {
 		pub CollectionSchema get(fn collection_schema): map hasher(blake2_128_concat) CollectionId => Option<NFTSchema>;
```
Comment (aliXsed) [inline crml/nft/src/lib.rs]: To me NFTSchema and all the code around that seems redundant. `Vec<NFTAttributeValues>` that we store for every token is powerful enough for encoding/decoding. It internally stores what we have named type ids.

File: crml/nft/src/lib.rs [hunk 3]
```diff
@@ -157,24 +156,33 @@ decl_storage! {
...
 		/// Map from collection to a base metadata URI for its token's offchain attributes
 		pub CollectionMetadataURI get(fn collection_metadata_uri): map hasher(blake2_128_concat) CollectionId => MetadataURI;
```
Comment (aliXsed) [inline crml/nft/src/lib.rs]: What's the use of this URI when we already have URL field for every token? Couldn't that URL be simply URI (as an optional part of a NFT)?

**Comment by aliXsed** [comment]
Also I am not sure creating node storage for the sake of simplifing some queries or easy access to some data is a good idea. If it's for the sake of generating usable RPCs, I would suggest we go for adding more custom RPCs.

File: crml/nft/src/lib.rs (hunk 3 above)
Comment (jordy25519) [reply crml/nft/src/lib.rs]: it's to save storage space, https://eips.ethereum.org/EIPS/eip-1155#metadata

**Comment by jordy25519** [comment]
run benchmark

File: crml/nft/src/lib.rs [hunk 4]
```diff
@@ -20,22 +20,34 @@
...
 //! Intended to be used "as is" by dapps and provide basic NFT feature set for smart contracts
 //! to extend.
+//!
+//! *Collections*:
+//!  A namespacing tool for logical grouping of related tokens
+//!  Tokens within a collection can have the same royalties fees, metadata base URIs, and owner address
+//!
+//! *Series*:
+//! A grouping of tokens within a collection namespace
+//! Tokens in the same series will have exact attributes and royalties.
```
Comment (miketamis) [inline crml/nft/src/lib.rs]: this implies that the series has control over the royalties. Maybe
"Tokens in the same series will have exact attributes and inherit it royalties info from the collection."

File: crml/nft/src/lib.rs [hunk 5]
```diff
@@ -214,208 +248,354 @@ decl_module! {
...
+		fn mint_unique(
+			origin,
+			collection_id: CollectionId,
+			owner: Option<T::AccountId>,
+			attributes: Vec<NFTAttributeValue>,
+			metadata_path: Option<Vec<u8>>,
+		) {
+			let series_id = Self::next_series_id(collection_id);
+			let _ = Self::mint_series(origin, collection_id, One::one(), owner, attributes, metadata_path)?;
+			IsSingleIssue::insert(collection_id, series_id, true);
```
Comment (miketamis) [inline crml/nft/src/lib.rs:315]: nice

File: crml/nft/src/tests.rs [hunk 6]
```diff
@@ -1581,15 +1403,252 @@ fn bid_fails_prechecks() {
...
+			collection_id,
+			None,
+			vec![],
+			None,
+		));
+
+		// One of one token
+		assert_noop!(
+			Nft::mint_additional(Some(collection_owner).into(), collection_id, series_id + 1, 5, None),
+			Error::<Test>::AddToSingleIssue
```
Comment (miketamis) [inline crml/nft/src/tests.rs:1651]: Instead of `AddToSingleIssue` maybe `AddToUniqueIssue`

**Review by miketamis — APPROVED** [review]
🃏🔥

**Pull request merged**